### ISSUE TITLE
Suggestion: change to issue labeling

### ISSUE THREAD

**Issue by jhux2** [trilinos/Trilinos#1619]
While trying to resolve a linker problem, I stumbled across the https://github.com/Unidata/netcdf-c/issues page.  As you can see, that project prefixes labels with categories.  I think Trilinos should consider adopting a similar format for labels.

```
package/blah
platform/blah
status/blah
type/blah
```

I believe this would help authors better categorize the issues, and make searching easier.

**Comment by jwillenbring** [comment]
@jhux2 How is this approach better than, for example 2 labels, package and blah, where you can search on both. This creates such an explosion of labels and doesn't it make it harder to search on just label blah?

**Comment by jhux2** [comment]
> How is this approach better than, for example 2 labels, package and blah, where you can search on both. This creates such an explosion of labels and doesn't it make it harder to search on just label blah?

@jwillenbring  Good questions.  I wasn't necessarily suggesting new labels.  Maybe a better way to view this is it would put the existing labels into logical categories.  In some ways, it's mimicking the categories that we had available in Bugzilla.  However, categorizing would probably make it easier to see labels that we should have but don't.

Here's one possible translation of the existing labels:
```
muelu  ==> packages/muelu
in progress ==> status/in progress
bug ==> type/bug
enhancement ==> type/enhancement
mac os ==> platform/mac os
wontfix ==> resolution/wontfix
```
In my opinion, this would make it easier to see what labels should be applied to an issue.  If you look at the label pulldown in the netcdf project I mentioned, the labels are nicely organized by category.

I don't think the categorization would complicate searching, either.  One could still search as usual, e.g., `label:packages/muelu label:status/in progress`.

**Comment by aprokop** [comment]
I agree with @jhux2. We could also fix the whole coloring issue here, assigning a specific color to each area (packages, status, platform, etc) instead of the random rainbow we have right now.

**Comment by aprokop** [comment]
Here is an example to match all current labels in Trilinos:

packages/*
  Amesos, Amesos2, ..., Framework, ...

type/*
  bug, build, documentation, enhancement, performance, process improvement, question, results impacting, retrospective, tests

status/*
  help wanted, in progress, in review, ready

resolved/*
  duplicate, invalid, wontfix

priority/*
  epic, story, task

system/*
  gpu, linux, manycore, windows

**Comment by jhux2** [comment]
@aprokop Small quibble: epic, story, and task don't belong under priority in my opinion.  I don't know what to call that category, but any of the three could be high, medium, etc. priority.

**Comment by aprokop** [comment]
@jhux2 agreed. scrum/*?

**Comment by jhux2** [comment]
> @jhux2 agreed. scrum/*?

Maybe.  what about `scope/*`.

**Comment by jwillenbring** [comment]
> I wasn't necessarily suggesting new labels.

I didn't realize that. I misunderstood what you were suggesting. It seems there would be some value then in this approach.

**Comment by srajama1** [comment]
Agree with @jhux2 . I don't care for a scrum label, but that is just me.

**Comment by crtrott** [comment]
You can actually rename existing labels so it will automatically effect everything.

**Comment by jhux2** [comment]
> You can actually rename existing labels so it will automatically effect everything.

@crtrott Good to know, thanks.  This question was asked at the TL meeting.

**Comment by aprokop** [comment]
Any decision on this? Do we want to move forward, or postpone the discussion till TUG?

**Comment by jhux2** [comment]
> Any decision on this? Do we want to move forward, or postpone the discussion till TUG?

@aprokop No decision -- we haven't had a Trilinos Leaders meeting in the last three weeks.

**Comment by maherou** [comment]
Sorry for the delay on this, due to canceled Trilinos Leaders meetings.

We can discuss next week (for sure).  In the meantime, since it is possible to change labels, and revert them if needed, it might be good to convert a few of them as an example.  Just a suggestion.


On Sep 13, 2017, at 4:04 PM, Jonathan Hu <notifications@github.com<mailto:notifications@github.com>> wrote:


Any decision on this? Do we want to move forward, or postpone the discussion till TUG?

@aprokop<https://github.com/aprokop> No decision -- we haven't had a Trilinos Leaders meeting in the last three weeks.

-

**Comment by bartlettroscoe** [comment]
I kind of like:

```
<category>: <val>
````

more than

```
<category>/<val>
```

Reads better in general I think (but just a preference)

**Comment by jhux2** [comment]
> Sorry for the delay on this, due to canceled Trilinos Leaders meetings.
> 
> We can discuss next week (for sure).  In the meantime, since it is possible to change labels, and revert them if needed, it might be good to convert a few of them as an example.  Just a suggestion.

I updated the labels `wontfix`, `duplicate`, `invalid` to `resolved/*`.

![newlabels-2](https://user-images.githubusercontent.com/12970120/30506954-ee6308d4-9a34-11e7-82a0-8c91d69469e5.png)

:+1: if you prefer this.

**Comment by jhux2** [comment]
> We can discuss next week (for sure).  In the meantime, since it is possible to change labels, and revert them if needed, it might be good to convert a few of them as an example.  Just a suggestion.

I updated the architecture labels to `system: *`.

![newlabels-1](https://user-images.githubusercontent.com/12970120/30507126-4c7d3812-9a36-11e7-852b-420fd9a50ade.png)

:+1: if you prefer this style.

**Comment by jhux2** [comment]
By the way, github has no way to do a wildcard search, e.g.,
`is:issue is:open label:"system: *"`
or
`is:issue is:open label:"resolved/*"`.

Just wanted to point this out.

**Comment by crtrott** [comment]
Maybe category/label is better because there is no white space. That may work better inside of github API calls for external tools.

**Comment by bartlettroscoe** [comment]
> Maybe category/label is better because there is no white space. That may work better inside of github API calls for external tools.

Looks like the GitHub API works fine with spaces in the labels.  See:

* https://stackoverflow.com/questions/41761927/github-api-issues-by-label-with-spaces

FYI: I just switched the github.com/TriBITSPub/TriBITS repo to using namspaced lables (e.g. `<category>: <value>`) and it works just fine and is quite readable.  Even waffe.io works fine (after updating the names of Kanban `stage` labels in the columns setting).

**Comment by jhux2** [comment]
I just updated the issue labels for `Amesos2`, `Ifpack2`, `Belos`, `MueLu`, `Tpetra`, `Xpetra`, and `Zoltan2` by prepending `package: ` (with a space after the colon).

**Comment by jhux2** [comment]
@trilinos/amesos2 @trilinos/ifpack2 @trilinos/belos @trilinos/muelu @trilinos/tpetra @trilinos/xpetra @trilinos/zoltan2

**Comment by bartlettroscoe** [comment]
> I just updated the issue labels for Amesos2, Ifpack2, Belos, MueLu, Tpetra, Xpetra, and Zoltan2 by prepending `packages: `(with a space after the colon).

@jhux2, this is is pretty long and makes for a long list of labels ugly.  Instead, can we consider removing `packages: ` and instead just providing label comments like I just did for the `Thyra` and the `Teuchos` labels?  With that, a tool tip pops up when you run your mouse pointer over it.  For example, for Thyra this says "Issues primarily dealing with the Thyra Package".

At the very least, can we shorten this to `pkg: Thyra`, `pkg: ROL`, `pkg: MueLu`?

At the very very least, can we shorten this from `packages: <pkg>` to `package: <pkg>`?  It makes no sense to say `packages: MueLu` because MueLu is not a "Packages", it a "Package".  For example, w have the label `system: linux` not `systems: linux`.

**Comment by jhux2** [comment]
@bartlettroscoe  No doubt, the string of labels are longer with the additional `package: ` prefix.

In my opinion, tool tips don't help organize labels, which was one of the original motivations for this issue.  See the link in the opening comment for an example of how nicely the labels are categorized.

Moving from `packages:` to `package:` makes sense.  I don't find `pkg: ` particularly pleasing, but I'm one person :).

**Comment by bartlettroscoe** [comment]
@jhux2 said:

> Moving from `packages: ` to `package: ` makes sense. I don't find `pkg:` particularly pleasing, but I'm one person :).

So let's just put `package: ` prefix on all of the Trilinos GitHub Issue Packages labels.  So unless there are any objections, I will just do it (It should not but 10 minutes so so).

**Comment by jhux2** [comment]
> So let's just put `package: ` prefix on all of the Trilinos GitHub Issue Packages labels. So unless there are any objections, I will just do it (It should not but 10 minutes so so).

@bartlettroscoe Go for it!

**Comment by bartlettroscoe** [comment]
Just a quick poll from Trilinos developers, please add emoji thumbs-up for:

* `pkg: <Package>`

or give thumbs-down for:

* `package: <Package>`

for the names of Trilinos GitHub labels for Trilinos packages.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe is there an emoji for "I'm indifferent but please pick one & stick to it"

@mhoemmen, yup, you just pick my preference which is thumbs-up :-)

**Comment by mhoemmen** [comment]
I'll go w/ @jmgate 's choice of "meh" ;-)

**Comment by bartlettroscoe** [comment]
Looks like the "thumbs-up" have it (3 to 1) so it will be `pkg: <Package>`.

**Comment by bartlettroscoe** [comment]
> Looks like as of 12/7/2018 the "thumbs-up" have it (4 to 1) so it will be `pkg: <Package>`

All done as shown at:

https://github.com/trilinos/Trilinos/labels?utf8=%E2%9C%93&q=pkg%3A+

I timed the work to do this from start to end and it took only 5m 55s.

**Comment by bartlettroscoe** [comment]
We might also consider standard prefixes for the following categories of labels:

* process type: 'epic', 'story', 'task' (pick 0 or 1 out of set)
* kanban stage: 'ready', 'in progress', 'in review' (pick 0 or 1 out of set)
* type: 'bug', 'enhancement', 'question', 'help wanted'  (pick 0 or 1 out of set) 
* impacting: 'build', 'test', 'documentation', 'performance'  (pick 0 or more out of set)
* program/customer: 'ATDM', 'Sierra', 'SPARC', 'EMPIRE'

I am just not sure what short prefix to use for these categories.  Ideas?

Then we have some undocumented or other odd-ball labels that I am not not sure to do with:
* 'Framework'? (What does this label really mean?)
* 'process improvement'?
* 'Release 12.14'?  (Should just delete this, no?)
* 'retrospective'?
* 'TeptraRF'?

All of the other labels have pretty clear documentation.

What say you?

**Comment by bartlettroscoe** [comment]
> Instead of kanban `stage`, what about `status`? 

@jhux2, sounds good to me.   I called this 'stage:' in the TriBITS repo shown at at:

* https://github.com/tribitspub/tribits/labels

How about:

* `stage`: `ready', `in progress`, `in review` (pick 0 or 1 out of set)

> `process type` doesn't really mean anything to me, either.

I don't know what to call this either.

NOTE: I was not suggesting the names of the exact prefixes, just trying to name the categories.

**Comment by jhux2** [comment]
@bartlettroscoe I did a little cleaning of the labels, using ideas from your comment [above](https://github.com/trilinos/Trilinos/issues/1619#issuecomment-445315595).  Please have a look.

`program/customer` is a bit long.  What do you think of the following?

`client: 'ATDM', 'Sierra', 'SPARC', 'EMPIRE', 'CASL', 'GEMMA'`

**Comment by bartlettroscoe** [comment]
> `client: 'ATDM', 'Sierra', 'SPARC', 'EMPIRE', 'CASL', 'GEMMA'`

@jhux2, I am okay with that.  I will just have to update a bunch of URL to GitHub queries (like I did when `bug` was changed to `type: bug`) but is okay.  Every label in the Trilinos GitHb project should be prefixed like this so I say go for it.

**Comment by jhux2** [comment]
@bartlettroscoe Change is made.  I'll quit fussing with the labels for a while (it's a nice concrete task when I want a break from something else :)).

**Comment by jhux2** [comment]
@bartlettroscoe wrote:
> Then we have some undocumented or other odd-ball labels that I am not not sure to do with:
> 
> * 'Framework'? (What does this label really mean?)

There are a lot more Framework labels than "PA: Framework".  I'd suggest
1. replace all instances of "PA: Framework" with "Framework"
2. delete "PA: Framework"
3. rename "Framework" to "PA: Framework"

> * 'process improvement'?

I know what it is, no idea how to handle it.
> * 'Release 12.14'?  (Should just delete this, no?)

One open issue is tracked by this.
> * 'retrospective'?

I believe this label flags issues that might have lessons-learned for process improvement
> * 'TeptraRF'?

used by Tpetra team to denote refactoring issues

**Comment by bartlettroscoe** [comment]
@jhux2,

Actually "PA: Framework" has an important and well-defined meeting for ATDM as described [here](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM-Specific+Trilinos+GitHub+Labels#ATDM-SpecificTrilinosGitHubLabels-LabelsforTrilinosProductAreas).  It is the label "Framework" that has no meaning currently.  If anything, keep "PA: Framework" and get rid of "Framework".

@jwillenbring, what say you?

**Comment by jwillenbring** [comment]
@bartlettroscoe @jhux2 

> Actually "PA: Framework" has an important and well-defined meeting for ATDM as described here. It is the label "Framework" that has no meaning currently. If anything, keep "PA: Framework" and get rid of "Framework".

Ideally I would like to keep framework for tasks the framework team is working on, but people tend to just throw the framework label on things, so that list of tasks isn't really controllable.

**Comment by jhux2** [comment]
> Ideally I would like to keep framework for tasks the framework team is working on, but people tend to just throw the framework label on things, so that list of tasks isn't really controllable.

@jwillenbring Going forward, what would be your preference for labeling new issues?  Do you think it would it be helpful to rename `Framework` to `Framework tasks` or something similar?

**Comment by jwillenbring** [comment]
@jhux2 

>Going forward, what would be your preference for labeling new issues? Do you think it would it be helpful to rename Framework to Framework tasks or something similar?

Worth a try I think. I will do that.

**Comment by jhux2** [comment]
Closing, labels have long since been reorganized as suggested in this issue.
